File: core/voxcore/clients/asr.py
"""ASR engine client: `/v1/audio/transcriptions`."""

import re
from pathlib import Path
from typing import Any, NamedTuple

from ..config import EngineCfg
from ..http import build_client, raise_for_status

# The engine leaks a language tag into the transcript -- once at the tail for
# Chinese, once per sentence for English.
LANG_TAG = re.compile(r"\s*<[a-z]{2}-[A-Z]{2}>")
FIRST_LANG = re.compile(r"<([a-z]{2})-[A-Z]{2}>")


class TranscriptionSegment(NamedTuple):
    start: float
    end: float
    text: str
    speaker: str | None = None
    id: str | int | None = None


class Transcription(NamedTuple):
    text: str
    lang: str | None
    duration: float | None = None
    segments: tuple[TranscriptionSegment, ...] | None = None


def parse_transcript(raw: str) -> Transcription:
    """Strip the leaked tags, and read the language from the first one.

    `verbose_json.language` is hardcoded to "en" by the engine and must not be used.
    The inline tag is the only real language detection available.
    """
    match = FIRST_LANG.search(raw)
    return Transcription(LANG_TAG.sub("", raw).strip(), match.group(1) if match else None)


def parse_response(payload: Any) -> Transcription:
    if not isinstance(payload, dict):
        return parse_transcript("")
    raw = payload.get("text", "")
    result = parse_transcript(raw if isinstance(raw, str) else "")
    parsed = []
    items = payload.get("segments")
    for item in items if isinstance(items, list) else []:
        if not isinstance(item, dict):
            continue
        if not isinstance(item.get("start"), (int, float)) or not isinstance(item.get("end"), (int, float)):
            continue
        if not isinstance(item.get("text"), str):
            continue
        parsed.append(TranscriptionSegment(
            start=float(item["start"]),
            end=float(item["end"]),
            text=item["text"],
            speaker=item.get("speaker") if isinstance(item.get("speaker"), str) else None,
            id=item.get("id") if isinstance(item.get("id"), (str, int)) else None,
        ))
    duration = payload.get("duration")
    return Transcription(
        result.text,
        result.lang,
        float(duration) if isinstance(duration, (int, float)) else None,
        tuple(parsed) if isinstance(items, list) else None,
    )


class ASRClient:
    def __init__(self, cfg: EngineCfg):
        self.cfg = cfg
        self._client = build_client(cfg)

    def close(self):
        self._client.close()

    def __enter__(self):
        return self

    def __exit__(self, *_):
        self.close()

    def transcribe(
        self,
        audio_path: str | Path,
        *,
        language: str = "auto",
        structured: bool = False,
        max_new_tokens: int | None = None,
    ) -> Transcription:
        path = Path(audio_path)
        with path.open("rb") as fh:
            data = {
                "model": self.cfg.model,
                "language": language,
                "response_format": "verbose_json" if structured else "json",
            }
            if max_new_tokens is not None:
                data["max_new_tokens"] = str(max_new_tokens)
            response = self._client.post(
                "/v1/audio/transcriptions",
                data=data,
                files={"file": (path.name, fh)},
            )
        return parse_response(raise_for_status(response).json())

File: packages/clients/src/asr.ts
import type {
  EngineConfig,
  Transcription,
  TranscriptionOptions,
  TranscriptionSegment,
} from "@voxstudio/contracts";
import { EngineClient, type Fetch } from "./http";
import { parseTranscript } from "./parsing";

export class AsrClient extends EngineClient {
  constructor(config: EngineConfig, fetch?: Fetch) {
    super(config, fetch);
  }

  async transcribe(
    audio: Blob,
    filename: string,
    language = "auto",
    options: TranscriptionOptions = {},
  ): Promise<Transcription> {
    const form = new FormData();
    form.set("model", this.config.model);
    form.set("language", language);
    form.set("response_format", options.responseFormat ?? "json");
    if (options.maxNewTokens !== undefined) {
      form.set("max_new_tokens", String(options.maxNewTokens));
    }
    form.set("file", audio, filename);
    const response = await this.request("/v1/audio/transcriptions", {
      method: "POST",
      body: form,
    });
    const payload: unknown = await response.json();
    const raw = typeof payload === "object" && payload !== null && "text" in payload
      && typeof payload.text === "string"
      ? payload.text
      : "";
    const transcription = parseTranscript(raw);
    if (typeof payload !== "object" || payload === null) return transcription;
    const record = payload as Record<string, unknown>;
    const segments = parseSegments(record.segments);
    const duration = typeof record.duration === "number" ? record.duration : undefined;
    return {
      ...transcription,
      ...(duration === undefined ? {} : { duration }),
      ...(segments === undefined ? {} : { segments }),
    };
  }
}

function parseSegments(value: unknown): TranscriptionSegment[] | undefined {
  if (!Array.isArray(value)) return undefined;
  const result: TranscriptionSegment[] = [];
  for (const item of value) {
    if (typeof item !== "object" || item === null) continue;
    const segment = item as Record<string, unknown>;
    if (
      typeof segment.start !== "number"
      || typeof segment.end !== "number"
      || typeof segment.text !== "string"
    ) continue;
    result.push({
      ...(typeof segment.id === "string" || typeof segment.id === "number" ? { id: segment.id } : {}),
      start: segment.start,
      end: segment.end,
      text: segment.text,
      ...(typeof segment.speaker === "string" ? { speaker: segment.speaker } : {}),
    });
  }
  return result;
}
